Projects

ActiveChooN opened the pull request:
<!---
Copyright (C) 2020 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

This PR adds ability to organize annotation tasks into projects

![изображение](https://user-images.githubusercontent.com/8168727/95076120-09ce6f00-071a-11eb-975e-6b24d852c337.png)

![изображение](https://user-images.githubusercontent.com/8168727/95076158-181c8b00-071a-11eb-86c6-7ff58a93a8e1.png)

partly related to #1278 

### List of TODOs:
- [x] Add advanced project selector to the create project page
- [x] Logger for projects
- [x] Enable search field for the projects page
- [x] Changelog, package versions, etc
- [x] Add tests for API

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

Manually

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
~~- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly~~
- [x] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

nmanovic commented:
@ActiveChooN , could you please update the PR and notify me? Just want to check the intermediate stage.

nmanovic commented:
@ActiveChooN , could you please update the PR and notify me? Just want to check the intermediate stage. I still see the bunch of merge conflicts.

ActiveChooN commented:
@nmanovic, guess it's mostly resolved.

nmanovic reviewed (changes requested):
Several comments from my side:
- The whole card should be clickable.
- Replace `Actions` by 3 vertical dots
- Submit button is located incorrectly (`create a new` task dialog)
![image](https://user-images.githubusercontent.com/40690625/97802201-ca306f80-1c52-11eb-9eb0-1a66bb6d3da3.png)
- Don't stretch thumbnail
![image](https://user-images.githubusercontent.com/40690625/97802276-52167980-1c53-11eb-9308-21284d5275bc.png)
- Something wrong with the menu item for projects
![image](https://user-images.githubusercontent.com/40690625/97802345-dd900a80-1c53-11eb-9200-ce88ed5dbba7.png)
- A logical error with notifications for a new project. How to reproduce: click `projects` tab, click `create a new project`. After that you will see the notification again.
- After I reload cvat page I got another bug. The project shows that I don't have tasks. Go to tasks page and I can see my new task. Go to projects page and it presents correct information.
![image](https://user-images.githubusercontent.com/40690625/97802551-44fa8a00-1c55-11eb-9953-9eedf4fd8513.png)
![image](https://user-images.githubusercontent.com/40690625/97802589-986cd800-1c55-11eb-9829-47b802cfd610.png)
![image](https://user-images.githubusercontent.com/40690625/97802597-a02c7c80-1c55-11eb-9e67-5b1141b8406b.png)

nmanovic commented:
@ActiveChooN , I see two major issues with the request:
- Need to send "formatting" PR in advance. Vitaly does it in the past but looks like we changed configuration files and need to run it again. Otherwise it is mostly impossible to review code because it is very difficult to distinguish a real change and refactoring. 
- Travis-ci build failed. Need to fix that.

bsekachev commented:
Hi, great patch! Here are some thoughts about how we could improve it:

![Screenshot from 2020-11-10 11-57-33](https://user-images.githubusercontent.com/40690378/98652299-7ad1f980-234c-11eb-8029-8030a5d8ba39.png)
1. I would remove this button and leave a hyperlink in task info block. Example:
**Project:** [Name of the project]()
Or at least instead of ID in the button text put the id in task info block. Also I would change "back to project" to  "go to a project" because a task could be opened from direct URL, or from tasks list, or from a job, and "back" is not applicable in these cases. But personally I more like the first solution with hyperlink in info block instead. 

2. Maybe just remove it? (project view doesn't have such the element)
3. Empty space when open a task from a project page. When open the task from tasks the space is left by "Created by ..."
4. Doesn't work for me. The same issue when I try to specify issue tracker or change task name
```

Request Method: | PATCH
-- | --
http://localhost:7000/api/v1/tasks/2
3.1.1
IntegrityError
UNIQUE constraint failed: engine_attributespec.id
```

And one question about labels. When create task within a project, there is a text "Project labels will be used". On a task page there aren't any text. Looks inconsistent. I would suggest two options here. The first one is to display all labels as usual but add readonly mode to the labels editor component. The second one is to remove labels from create task page also. But task page in this case looks really empty. 

Questions about the issue tracker. Should we allow a user to setup the task issue tracker differ of project issue tracker? Should we use a project issue tracker as default value for the task issue tracker as default if available?

Wrong position of the submit button:
![Screenshot from 2020-11-10 11-49-55](https://user-images.githubusercontent.com/40690378/98654788-cf2aa880-234f-11eb-99f1-b8cf9810d5a8.png)

Project actions, task actions. Let's do their appearance consistent.
![Screenshot from 2020-11-10 12-27-18](https://user-images.githubusercontent.com/40690378/98655112-2b8dc800-2350-11eb-9fc6-20661e4c828a.png)
![Screenshot from 2020-11-10 12-27-31](https://user-images.githubusercontent.com/40690378/98655117-2cbef500-2350-11eb-90d9-f2e973648934.png)
 
The header looks messy. How about adding a preview and some additional offsets? (for example between created by and issue tracker)
![Screenshot from 2020-11-10 12-28-39](https://user-images.githubusercontent.com/40690378/98655231-4a8c5a00-2350-11eb-8912-4beef4cea2f5.png)

bsekachev commented on tests/cypress/support/index.js:
```diff
@@ -6,7 +6,7 @@ require('./commands');
 
 before(() => {
-    if (Cypress.browser.name === 'firefox') {
+    if (Cypress.browser.family !== 'chromium') {
```
Is this your change?

ActiveChooN replied:
I've merged PR from @dvkruchinin with some UI test fixes.

bsekachev commented:
@ActiveChooN 
The PR has a lot of changed files in directory ``tests/cypress``.  Could you please suggest why? Is it really necessary?

bsekachev commented:
Could you please get this PR rid of minor refactoring? It is difficult to review the PR
A lot of files without any significant changes, only with code-style refactoring (apparently performed by prettier/eslint), but prettier & eslint runs only for changed files and it is not clear why these files are formatted.

bsekachev commented on cvat-ui/src/reducers/interfaces.ts:
```diff
@@ -21,6 +21,37 @@ export interface AuthState {
+    [key: string]: string | number | null | undefined;
+}
+
+export interface Project {
+    instance: any; // cvat-core instance
+}
```
Do we really need this additional instance with a single field? Such the instance has been created for Task because there was a preview also.

bsekachev commented on cvat-ui/src/components/create-task-page/create-task-page.tsx:
```diff
@@ -20,8 +21,16 @@ interface Props {
-    const { error, status, taskId, onCreate, installedGit } = props;
+export default function CreateTaskPage(props: Props & RouteComponentProps): JSX.Element {
+    const {
+        error, status, taskId, location, onCreate, installedGit,
```
Let's use ``useLocation`` hook instead of wrapping the component with ``withRouter``

bsekachev commented on cvat-ui/src/containers/create-task-page/create-task-page.tsx:
```diff
@@ -34,4 +35,4 @@ function mapStateToProps(state: CombinedState): StateToProps {
 }
 
-export default connect(mapStateToProps, mapDispatchToProps)(CreateTaskComponent);
+export default connect(mapStateToProps, mapDispatchToProps)(withRouter(CreateTaskComponent));
```
Let's use hook inside the component instead

bsekachev commented on cvat-ui/src/containers/annotation-page/standard-workspace/controls-side-bar/setup-tag-popover.tsx:
```diff
@@ -7,6 +7,7 @@ import { connect } from 'react-redux';
 
 import { CombinedState, ObjectType } from 'reducers/interfaces';
 import { createAnnotationsAsync, rememberObject } from 'actions/annotation-actions';
+// eslint-disable-next-line max-len
```
In develop for cvat-ui we use ``'max-len': ['error', { code: 120, ignoreStrings: true }],``, so, I am not sure we need this command here

bsekachev commented on cvat-ui/src/containers/annotation-page/standard-workspace/controls-side-bar/draw-shape-popover.tsx:
```diff
@@ -9,6 +9,7 @@ import { RadioChangeEvent } from 'antd/lib/radio';
 import { CombinedState, ShapeType, ObjectType } from 'reducers/interfaces';
 import { rememberObject } from 'actions/annotation-actions';
 import { Canvas, RectDrawingMethod, CuboidDrawingMethod } from 'cvat-canvas-wrapper';
+// eslint-disable-next-line max-len
```
In develop for cvat-ui we use 'max-len': ['error', { code: 120, ignoreStrings: true }],, so, I am not sure we need this command here

bsekachev commented on cvat-ui/src/containers/annotation-page/standard-workspace/controls-side-bar/controls-side-bar.tsx:
```diff
@@ -16,6 +16,7 @@ import {
     pasteShapeAsync,
     resetAnnotationsGroup,
 } from 'actions/annotation-actions';
+// eslint-disable-next-line max-len
```
Please remove this and other such the commands

bsekachev commented on cvat-ui/src/components/tasks-page/empty-list.tsx:
```diff
@@ -31,6 +31,10 @@ export default function EmptyListComponent(): JSX.Element {
             <Row type='flex' justify='center' align='middle'>
                 <Col>
                     <Link to='/tasks/create'>create a new task</Link>
+                    &nbsp;
```
Let's remove it ```&nbsp;```
If you need the text on different lines, try to wrap the lines in <div> for example, or use <br />

bsekachev commented on cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx:
```diff
@@ -22,7 +22,7 @@ type ResetPasswordConfirmFormProps = {
     onSubmit(resetPasswordConfirmData: ResetPasswordConfirmData): void;
 } & FormComponentProps &
-    RouteComponentProps;
+RouteComponentProps;
```
Let's put it on the previous line. Looks strange.

bsekachev commented on cvat-ui/src/components/projects-page/styles.scss:
```diff
@@ -0,0 +1,119 @@
+@import '../../base.scss';
+
+.cvat-projects-page {
+    padding-top: 15px;
```
Please, in all scss files in cvat-ui use $grid-unit-size variable with a multiplier instead of absolute number in pixels. It is related with current design collaboration.
For example in this case it should be: ``padding-top: $grid-unit-size * 2`` is the nearest value to 15.

bsekachev commented on cvat-ui/src/components/project-page/project-page.tsx:
```diff
@@ -0,0 +1,106 @@
+}
+
+export default function ProjectPageComponent(): JSX.Element {
+    // TODO: need to optimize renders here
```
Is this TODO performed?

ActiveChooN replied:
Yes, forgot to delete notification

bsekachev commented on cvat-canvas/src/typescript/canvasModel.ts:
```diff
@@ -288,15 +288,23 @@ export class CanvasModelImpl extends MasterImpl implements CanvasModel {
-                mutiplier * ((x - this.data.imageSize.width / 2) * (oldScale / this.data.scale - 1)) * this.data.scale;
-            this.data.left -=
-                mutiplier * ((y - this.data.imageSize.height / 2) * (oldScale / this.data.scale - 1)) * this.data.scale;
+            // prettier-ignore
```
Please, revert changes in files where aren't any significant changes, like this.

ActiveChooN replied:
As I mentioned this changes was made by prettier due to merge conflicts in merge. Also without this comments prettier and eslint was in conflict, so i added this comment.

bsekachev replied:
prettier & eslint are run only for git staged files. 
If you haven't changed them, they aren't linted.

bsekachev replied:
Please, try to checkout unchanged files to develop and commit them with --no-verify.
Example:
```
git checkout develop -- cvat-canvas/src/typescript/canvasModel.ts
git commit -m "test" --no-verify
```

bsekachev commented on cvat-core/.eslintrc.js:
```diff
@@ -4,6 +4,7 @@
 
 module.exports = {
     env: {
+        amd: true,
```
Why do you need this change?

ActiveChooN replied:
Due to require warining

bsekachev replied:
Set ``node`` (row below) to true instead

bsekachev commented:
A list of files where changes aren't related with the PR. It may be not the full list and I maybe wrong in some of items. Need an additional check. 

cvat-canvas/src/typescript/autoborderHandler.ts 
cvat-canvas/src/typescript/canvasModel.ts
cvat-canvas/src/typescript/canvasView.ts 
cvat-canvas/src/typescript/consts.ts
cvat-canvas/src/typescript/cuboid.ts 
cvat-canvas/src/typescript/drawHandler.ts
cvat-canvas/src/typescript/editHandler.ts
cvat-canvas/src/typescript/groupHandler.ts
cvat-canvas/src/typescript/interactionHandler.ts
cvat-canvas/src/typescript/mergeHandler.ts 
cvat-canvas/src/typescript/shared.ts 
cvat-canvas/src/typescript/zoomHandler.ts 
cvat-ui/src/actions/annotation-actions.ts 
cvat-ui/src/actions/auth-actions.ts 
cvat-ui/src/actions/boundaries-actions.ts 
cvat-ui/src/actions/formats-actions.ts 
cvat-ui/src/actions/models-actions.ts 
cvat-ui/src/actions/share-actions.ts 
cvat-ui/src/actions/useragreements-actions.ts
cvat-ui/src/components/actions-menu/dump-submenu.tsx
cvat-ui/src/components/actions-menu/load-submenu.tsx
cvat-ui/src/components/annotation-page/annotation-page.tsx
cvat-ui/src/components/annotation-page/appearance-block.tsx
cvat-ui/src/components/annotation-page/attribute-annotation-workspace/attribute-annotation-sidebar/attribute-editor.tsx
cvat-ui/src/components/annotation-page/attribute-annotation-workspace/attribute-annotation-sidebar/attribute-switcher.tsx
cvat-ui/src/components/annotation-page/attribute-annotation-workspace/attribute-annotation-sidebar/object-switcher.tsx
cvat-ui/src/components/annotation-page/standard-workspace/canvas-point-context-menu.tsx
cvat-ui/src/components/annotation-page/standard-workspace/canvas-wrapper.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/cursor-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-cuboid-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-points-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-polygon-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-polyline-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-rectangle-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/group-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/merge-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/move-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/resize-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/rotate-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/setup-tag-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/setup-tag-popover.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/split-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx
cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/color-picker.tsx
cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/object-item-attribute.tsx
cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/object-item-details.tsx
cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/object-item-menu.tsx
cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/object-item.tsx
cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/objects-side-bar.tsx
cvat-ui/src/components/annotation-page/standard-workspace/standard-workspace.tsx
cvat-ui/src/components/annotation-page/top-bar/annotation-menu.tsx
cvat-ui/src/components/annotation-page/top-bar/left-group.tsx
cvat-ui/src/components/annotation-page/top-bar/player-buttons.tsx
cvat-ui/src/components/change-password-modal/change-password-modal.tsx
cvat-ui/src/components/labels-editor/raw-viewer.tsx 
cvat-ui/src/components/login-page/cookie-policy-drawer.tsx
cvat-ui/src/components/model-runner-modal/model-runner-dialog.tsx
cvat-ui/src/components/models-page/models-page.tsx 
cvat-ui/src/components/tasks-page/task-list.tsx
cvat-ui/src/components/tasks-page/tasks-page.tsx
cvat-ui/src/components/tasks-page/top-bar.tsx
cvat-ui/src/containers/actions-menu/actions-menu.tsx
cvat-ui/src/containers/annotation-page/standard-workspace/canvas-context-menu.tsx
cvat-ui/src/containers/annotation-page/standard-workspace/canvas-wrapper.tsx 
cvat-ui/src/containers/annotation-page/standard-workspace/objects-side-bar/object-buttons.tsx
cvat-ui/src/containers/annotation-page/standard-workspace/propagate-confirm.tsx
cvat-ui/src/containers/annotation-page/top-bar/statistics-modal.tsx
cvat-ui/src/containers/annotation-page/top-bar/top-bar.tsx
cvat-ui/src/containers/models-page/models-page.tsx
cvat-ui/src/containers/task-page/details.tsx
cvat-ui/src/containers/tasks-page/tasks-page.tsx
cvat-ui/src/cvat-store.ts
cvat-ui/src/reducers/annotation-reducer.ts
cvat-ui/src/reducers/auth-reducer.ts
cvat-ui/src/reducers/settings-reducer.ts
cvat-ui/src/reducers/tasks-reducer.ts

azhavoro commented on cvat/apps/engine/serializers.py:
```diff
@@ -44,6 +44,47 @@ class Meta:
+            logger.info("{} label was updated".format(db_label.name))
+        if not validated_data.get('color', None):
+            label_names = [l.name for l in
+                models.Label.objects.filter(**instance).exclude(id=db_label.id).order_by('id')
```
```suggestion
                instance.label_set.exclude(id=db_label.id).order_by('id')
```

ActiveChooN commented:
@bsekachev 

>
>
> Also I would change "back to project" to "go to a project" because a task could be opened from direct URL, or from tasks list, or from a job, and "back" is not applicable in these cases.

I guess "Back to project" is more convenient in case website navigation. It’s not about browsing history. Also maybe it’s a good idea to add some breadcrumbs for navigation instead of it. For example: `/ project / 1 / task / 1`. In this case every word would have link to the correspondent page.

>
>
> Maybe just remove it? (project view doesn't have such the element)

Sure

>
>
> Empty space when open a task from a project page. When open the task from tasks the space is left by "Created by ..."

Fixed

>
>
> Doesn't work for me. The same issue when I try to specify issue tracker or change task name

Can you share json dump of the project, please? Cannot reproduce the issue.

>
>
> And one question about labels.

I guess we can use some form of small readonly labeleditor, but in my opinion it is task for separate PR. Can you create related issue?

>
>
> Questions about the issue tracker.

I guess we should not restrict users in such way. But it is discussible question about task issue tracker default from project issue tracker. I guess we can add this if user ever will ask us.

>
>
> Project actions, task actions. Let's do their appearance consistent.

Fixed

> 
>
> The header looks messy. How about adding a preview and some additional offsets? (for example between created by and issue tracker)

Fixed

>
>
> Could you please get this PR rid of minor refactoring?

Tried to remove all

bsekachev commented on cvat-ui/src/components/tasks-page/empty-list.tsx:
```diff
@@ -31,6 +31,10 @@ export default function EmptyListComponent(): JSX.Element {
             <Row type='flex' justify='center' align='middle'>
                 <Col>
                     <Link to='/tasks/create'>create a new task</Link>
+                    <br />
+                    <Text type='secondary'>or try to</Text>
+                    <br />
+                    <Link to='/projects/create'>create a new project</Link>
```
It seems I was wrong about <br /> because understood wrong the previous solution.
Instead of <br /> we just need a couple of additional spaces:

Have you tried something like that?
```
<Link to='/tasks/create'>create a new task</Link>
<Text type='secondary'> or try to </Text>
<Link to='/projects/create'>create a new project</Link>
```

ActiveChooN replied:
Fixed

bsekachev commented:
> It’s not about browsing history. 

Exactly. And it's why better to use "Go" instead  of "Back"/"Forward"

> breadcrumbs for navigation instead of it.

In my opinion browser URL could perform the role of such the breadcrumb. But in current URL scheme, yes we need something like that. But probably in a different PR. 

> Cannot reproduce the issue.

Also cannot reproduce it anymore. I will comment in advanced if I meet the issue.

> Can you create related issue? 

Please, submit the issue on your own

bsekachev commented:
An additional issue I came around.
When I open /tasks page (this task has been created before migration if it might be important):
![Screenshot from 2020-11-13 11-15-09](https://user-images.githubusercontent.com/40690378/99045000-b5809f80-25a1-11eb-9385-c7ccd130e799.png)

And what is about default redirecting? We discussed it with the team and decided to open /tasks by default, is not it?

bsekachev commented:
When I open a page with a project:
![Screenshot from 2020-11-13 11-19-15](https://user-images.githubusercontent.com/40690378/99045242-160fdc80-25a2-11eb-836c-0d162f4163b4.png)

bsekachev commented:
I was able to submit a project without any labels. Is it expected?

bsekachev commented:
When project selected on create/task page the button "Submit" is still in the wrong place. 
![Screenshot from 2020-11-13 11-24-48](https://user-images.githubusercontent.com/40690378/99045729-d990b080-25a2-11eb-868d-4748bca76be8.png)

ActiveChooN commented:
@bsekachev,

>
>
> In my opinion browser URL could perform the role of such the breadcrumb.

As i remember in newer version of chrome full URL might be hidden and only domain is shown.

>
>
> I was able to submit a project without any labels. Is it expected?

Yes, because  you can create a project and then move task into it and labels will be copied from the task.

bsekachev commented:
I was able to create a task without any labels (first created a project, then a task within the project) and now I cannot open it (what is actually expected). I suppose if a project doesn't have any labels, we should specify them in the first created task (or they are automatically copied from the first moved task according to your intentions)

bsekachev commented on cvat-core/src/api-implementation.js:
```diff
@@ -184,8 +190,12 @@
                 }
             }
 
+            if ('projectId' in filter && Object.keys(filter).length > 1) {
+                throw new ArgumentError('Do not use the filter field "projectId" with other');
+            }
```
Couldn't this field be used with ``page``?

ActiveChooN replied:
In general you are right, it was added in case of usage for only getting all project tasks.

bsekachev commented on cvat-core/src/api-implementation.js:
```diff
@@ -199,11 +209,48 @@
+            }
+
+            const searchParams = new URLSearchParams();
+            // TODO: need to check search fields
```
What to you mean with this TODO? The fields looks to be checked with checkFilter & additional conditions above

ActiveChooN replied:
Forgot to check TODO was deleted

bsekachev commented on cvat-core/src/api-implementation.js:
```diff
@@ -199,11 +209,48 @@
+            return projects;
         };
 
+        cvat.projects.searchNames.implementation = async (search, limit) => serverProxy.projects.searchProjectNames(search, limit);
```
Why not implement it with ``get`` (according to users, just provide BasicProjectSerializer that includes name and id)?

ActiveChooN replied:
It was done like you say, but I used different method because `get` method returns `Project` object and this method returns only `{id: ..., name: ...}` structure

bsekachev commented on cvat-core/src/server-proxy.js:
```diff
@@ -827,6 +908,17 @@
 
+                    projects: {
+                        value: Object.freeze({
+                            getProjects,
+                            searchProjectNames,
+                            saveProject,
+                            createProject,
+                            deleteProject,
```
Just suggestion:
```js
Object.freeze({
   get: getProjects,
   search: searchProjectNames,
   save: saveProject,
   create: createProject,
   delete: deleteProject
})
```

bsekachev commented on cvat-core/src/session.js:
```diff
@@ -951,6 +952,16 @@
                     },
+                    /**
+                     * @name projectId
+                     * @type {integer}
```
Suppose it might be null

bsekachev commented on cvat-core/src/session.js:
```diff
@@ -1406,9 +1417,234 @@
+     * @memberof module:API.cvat.classes
+     * @extends Session
+     */
+    class Project extends Session {
```
Session is supposed to be have quite wide API:  annotations, actions, frames, logger, etc.
Project doesn't provide it, so, I think we should not extends the Session class here

ActiveChooN replied:
Sure, moved to separate file

bsekachev commented on cvat-core/src/session.js:
```diff
@@ -1406,9 +1417,234 @@
+                    labels: {
+                        get: () => [...data.labels],
+                        set: (labels) => {
+                            if (!Array.isArray(labels)) {
+                                throw new ArgumentError('Value must be an array of Labels');
+                            }
+
+                            for (const label of labels) {
+                                if (!(label instanceof Label)) {
+                                    throw new ArgumentError(
+                                        `Each array value must be an instance of Label. ${typeof label} was found`,
+                                    );
+                                }
+                            }
```
```if (!Array.isArray(labels) || labels.some((label) => !(label instanceof Label)))```

bsekachev commented on cvat-core/src/session.js:
```diff
@@ -1406,9 +1417,234 @@
+         * @method save
+         * @returns {module:API.cvat.classes.Project}
+         * @memberof module:API.cvat.classes.Project
+         * been created yet. It called in order to notify about creation status.
+         * It receives the string parameter which is a status message
```
Extra lines?

bsekachev commented:
Why cannot we specify bug tracker during creating a project? A task has such the option. Is it planned to be implemented?

bsekachev commented on cvat-ui/src/components/create-task-page/project-search-field.tsx:
```diff
@@ -0,0 +1,91 @@
+        if (value && !projects.filter((project) => project.id === value).length) {
+            core.projects.get({ id: value }).then((result: Project[]) => {
+                const [project] = result;
+                setProjects([...projects, {
+                    id: project.id,
+                    name: project.name,
+                }]);
```
Why not ```setProjects([...projects, value]);```?

ActiveChooN replied:
In projects state we have such structure `{id: ..., name: ...}` wich was returned by `searchNames` method, so we keep that for data consistency

bsekachev commented on cvat-ui/src/components/projects-page/styles.scss:
```diff
@@ -0,0 +1,119 @@
+        display: flex;
+
+        > span:nth-child(2) {
+            width: 200px;
```
Please, do not use raw pixels

bsekachev commented on cvat-ui/src/components/projects-page/styles.scss:
```diff
@@ -0,0 +1,119 @@
+
+    img {
+        height: 100%;
+        max-height: 146px;
```
And here

bsekachev commented:
@azhavoro Hi, Could you please look at python files and describe your opinion?

dvkruchinin commented:
Hello.
At the moment, I am finishing writing tests for the specified scenario. And faced the problem of determining the system user who created the project. 
![cvat_projects](https://user-images.githubusercontent.com/33020454/99492405-9194d380-297e-11eb-994d-b52a6fb74bae.JPG)
Also with the problem of assigning the project to another user. After the assignment and login of the second user, I see the task that was created in the project (on the task list page). But on the project list page, the project is missing.

ActiveChooN commented on cvat-ui/src/actions/tasks-actions.ts:
```diff
@@ -381,6 +367,9 @@ export function createTaskAsync(data: any): ThunkAction<Promise<void>, {}, {}, A
             use_cache: data.advanced.useCache,
         };
 
+        if (data.projectId) {
```
By default database pk is >0, so i guess it will be enough here.

ActiveChooN commented on cvat-ui/src/components/create-project-page/create-project-content.tsx:
```diff
@@ -0,0 +1,112 @@
+            });
+        }
+
+        if (!projectName) return;
```
Guess it's quite the same thing

ActiveChooN commented on cvat-core/src/api-implementation.js:
```diff
@@ -163,6 +168,7 @@
         cvat.tasks.get.implementation = async (filter) => {
             checkFilter(filter, {
                 page: isInteger,
+                projectId: isInteger,
```
Done

ActiveChooN commented on cvat-ui/src/actions/projects-actions.ts:
```diff
@@ -0,0 +1,277 @@
+// Copyright (C) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+
```
Done

ActiveChooN commented:
@bsekachev 

> 
> 
> I was able to create a task without any labels (first created a project, then a task within the project) and now I cannot open it (what is actually expected). I suppose if a project doesn't have any labels, we should specify them in the first created task (or they are automatically copied from the first moved task according to your intentions)
> 
> UPD: Just saw that if after that you update labels in the project, the task could be opened. That's great, but not very obvious. At least I think need to add a relevant message during job loading (now the message looks like "cannot read property 'id' of undefined", what doesn't clarify the root cause.

Agree with you, added custom error with clearer description. Will it be enough in your opinion?



> 
> 
> Why cannot we specify bug tracker during creating a project? A task has such the option. Is it planned to be implemented?

Sure. Added

@dvkruchinin 

> And faced the problem of determining the system user who created the project.

It was a bug, should be fixed now.

bsekachev commented:
> Agree with you, added custom error with clearer description. Will it be enough in your opinion?

For the initial version I think it is enough.

bsekachev commented on cvat-ui/src/actions/projects-actions.ts:
```diff
@@ -0,0 +1,308 @@
+
+export function getProjectsAsync(query: Partial<ProjectsQuery>): ThunkAction {
+    return async (dispatch: ActionCreator<Dispatch>): Promise<void> => {
+        dispatch(getProjects());
```
dispatch(projectsActions.getProjects())

bsekachev replied:
The same comment is applicable to all dispatches in this file

ActiveChooN replied:
@bsekachev, done

bsekachev commented on cvat-ui/src/actions/projects-actions.ts:
```diff
@@ -0,0 +1,308 @@
+    return action;
+}
+
+function getProjects(): AnyAction {
+    const action = {
+        type: ProjectsActionTypes.GET_PROJECTS,
+        payload: {},
+    };
+
+    return action;
+}
```
Probably should be removed as other duplicated action creators below.

nmanovic commented:
```
FAIL: test_api_v1_projects_id_tasks_user (engine.tests.test_rest_api.ProjectListOfTasksAPITestCase)
4204----------------------------------------------------------------------
4205Traceback (most recent call last):
4206  File "/home/django/cvat/apps/engine/tests/test_rest_api.py", line 1012, in test_api_v1_projects_id_tasks_user
4207    self.assertListEqual(
4208AssertionError: Lists differ: ['my multijob task', 'super task'] != ['my multijob task', 'my task #1', 'my task #2', 'super task']
```

ActiveChooN commented:
> 
> 
> ```
> FAIL: test_api_v1_projects_id_tasks_user (engine.tests.test_rest_api.ProjectListOfTasksAPITestCase)
> 4204----------------------------------------------------------------------
> 4205Traceback (most recent call last):
> 4206  File "/home/django/cvat/apps/engine/tests/test_rest_api.py", line 1012, in test_api_v1_projects_id_tasks_user
> 4207    self.assertListEqual(
> 4208AssertionError: Lists differ: ['my multijob task', 'super task'] != ['my multijob task', 'my task #1', 'my task #2', 'super task']
> ```

@nmanovic, fixed

jahaniam commented:
I believe automatic annotation for tasks that they are assigned to a project is failing. They show successfully done but the result is not being saved/showed.

bsekachev commented:
@jahaniam , @ActiveChooN 

#2644 may be related
